src/compas/geometry/xforms/quaternion.py

```diff
@@ -0,0 +1,225 @@
+    :math:`p = rq`,
+    is also unit-length.
+
+
+    References
+    ----------
+    .. [1] http://mathworld.wolfram.com/Quaternion.html
+    .. [2] http://mathworld.wolfram.com/HamiltonsRules.html
+    .. [3] https://github.com/matthew-brett/transforms3d/blob/master/transforms3d/quaternions.py
+
+    """
+
+    def __init__(self, w, x, y, z):
+
+        self.w = float(w)
+        self.x = float(x)
+        self.y = float(y)
+        self.z = float(z)
+
+    def __iter__(self):
+        return iter([self.w, self.x, self.y, self.z])
+
+    def __str__(self):
+        return "Quaternion = %s" % list(self)
+
+    def __repr__(self):
+        return 'Quaternion({:.{prec}f}, {:.{prec}f}, {:.{prec}f}, {:.{prec}f})'.format(*self, prec=6)
+
+    def __mul__(self, other):
+        """Multiply operator for two quaternions.
+
+        Parameters
+        ----------
+        other
+            A Quaternion object.
+
+        Returns
+        -------
+        Quaternion
+            The product P = R * Q of this quaternion (R) multiplied by other quaternion (Q).
+
+        Examples
+        --------
+        >>> Q = Quaternion(1.0, 1.0, 1.0, 1.0).unitized()
+        >>> R = Quaternion(0.0,-0.1, 0.2,-0.3).unitized()
+        >>> P = R*Q
+        >>> P.is_unit
+        True
+
+        Notes
+        -----
+        Multiplication of two quaternions R*Q can be interpreted as applying rotation R to an orientation Q,
+        provided that both R and Q are unit-length.
+        The result is also unit-length.
+        Multiplication of quaternions is not commutative!
+        """
+
+        p = quaternion_multiply(list(self), list(other))
+        return Quaternion(*p)
+
+    @classmethod
+    def from_frame(cls, frame):
+        """Creates a ``Quaternion`` object from a ``Frame`` object.
+
+        Parameters
+        ----------
+        frame : :obj:`Frame`
+
+        Returns
+        -------
+        :obj:`Quaternion`
+            The new constructed ``Quaternion`` object.
+
+        Example
+        -------
+        >>> from compas.geometry import Frame
+        >>> q = [1., -2., 3., -4.]
+        >>> F = Frame.from_quaternion(q)
+        >>> Q = Quaternion.from_frame(F)
+        >>> allclose(list(Q.canonized()), quaternion_canonize(quaternion_unitize(q)))
+        True
+        """
+
+        w, x, y, z = frame.quaternion
+        return cls(w, x, y, z)
+
+    @property
+    def wxyz(self):
+        """
+        Returns the quaternion as a list of float in the "wxyz" convention.
+        """
+        return [self.w, self.x, self.y, self.z]
+
+    @property
+    def xyzw(self):
+        """
+        Returns the quaternion as a list of float in the "xyzw" convention.
+        """
+        return [self.x, self.y, self.z, self.w]
+
+    @property
+    def norm(self):
+        """
+        Returns the length (euclidean norm) of the quaternion.
+        """
+        return quaternion_norm(list(self))
```

Comment:
the conversion to list seems unnecessary because the class implements the iterator protocol